### PR TITLE
Added application Readout Error Mitigation

### LINKED ISSUE #1642: Readout Error Mitigation
### Required prerequisites

- [X] Search the [issue tracker](https://github.com/NVIDIA/cuda-quantum/issues) to check if your feature has already been mentioned or rejected in other issues.

### Describe the feature

# Background
Readout errors are caused by imperfect qubit measurement, and are a common source of error in quantum computing. Properly modeling these errors in simulation can give the user tools to better understand how these errors and how to mitigate them when running on actual quantum devices. 

There is a large body of work studying readout error, particularly in the context of how to mitigate it [[1](https://mitiq.readthedocs.io/en/stable/guide/rem-5-theory.html)], [[2](https://mitiq.readthedocs.io/en/stable/guide/rem-3-options.html)].

# Description
Develop a CUDA-Q application and apply readout error to the measurement results of a sampled CUDA-Q kernel. Next, perform readout error mitigation (REM) on those results. This application can be written in either C++ or Python.

The implementation should include two versions of both the measurement error and the subsequent measurement error mitigation: independent qubit measurement error model and full confusion matrix inversion.

# References
[1] https://mitiq.readthedocs.io/en/stable/guide/rem-5-theory.html
[2] https://mitiq.readthedocs.io/en/stable/guide/rem-3-options.html

### PR DESCRIPTION
**Pull request by Gopal-Dahale**
## Purpose

Added Readout Error Mitigation (REM) notebook in `examples/python/tutorials`.

## Description

The tutorial covers REM and three approaches to it:

- Using a  single qubit model
- using $k$ local confusion matrices
- A full confusion matrix for each $2^n$ combination

Resolves #1642 

The notebook took inspiration from:
1. [Mitiq REM theory](https://mitiq.readthedocs.io/en/stable/guide/rem-5-theory.html)
2. [Mitiq REM options](https://mitiq.readthedocs.io/en/stable/guide/rem-3-options.html)
3. [Qiskit experiments](https://qiskit-extensions.github.io/qiskit-experiments/manuals/measurement/readout_mitigation.html)
4. [Qiskit textbook, Measurement Error Mitigation](https://github.com/qiskit-community/qiskit-textbook/blob/main/content/ch-quantum-hardware/measurement-error-mitigation.ipynb)

## Mentions
@justinlietz

### CODE REVIEW

**Comment by Gopal-Dahale** [comment]
I have read the Contributor License Agreement and I hereby accept the Terms.

**Comment by Gopal-Dahale** [comment]
recheck

**Comment by Gopal-Dahale** [comment]
@justinlietz let me know your thoughts.

File: docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb
Comment (1tnguyen) [inline docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb:12]: Typo: "psuedoinverse"?

File: docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb
Comment (1tnguyen) [inline docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb:439]:
```suggestion
    "mitigated = np.array(np.dot(A_pinv, noisy_res_1), dtype=int)  # Obtain mitigated counts\n",
```

File: docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb
Comment (1tnguyen) [inline docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb:38]: If extra packages are needed, could you please add `pip install` session, like in this notebook?https://github.com/NVIDIA/cuda-quantum/blob/main/docs/sphinx/examples/python/tutorials/hybrid_qnns.ipynb

File: docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb
Comment (1tnguyen) [inline docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb:71]: Could you please add a docstring for this function?

File: docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb
Comment (1tnguyen) [inline docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb:188]: Could you please add a docstring for this function?

File: docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb
Comment (1tnguyen) [inline docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb:219]: Could you please add a docstring for this function?

File: docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb
Comment (Gopal-Dahale) [reply docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb:12]: Yes. I think it should be "pseudoinverse". The [mitiq](https://mitiq.readthedocs.io/en/stable/guide/rem-5-theory.html) docs also mistyped it.

File: docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb
Comment (1tnguyen) [inline docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb:488]: Could you please add some description text for this `print(mitigated)` and the other below (corrected for negative counts) so that we can understand the output easily?

File: docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb
Comment (1tnguyen) [inline docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb:667]: same as https://github.com/NVIDIA/cuda-quantum/pull/1765/files#r1634250902

File: docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb
Comment (1tnguyen) [inline docs/sphinx/examples/python/tutorials/readout_error_mitigation.ipynb:830]: Same as https://github.com/NVIDIA/cuda-quantum/pull/1765/files#r1634250902

**Comment by Gopal-Dahale** [comment]
@1tnguyen I have updated the notebook as per the comments. Can you take a look?

**Comment by 1tnguyen** [comment]
Hi @Gopal-Dahale 

Thanks for the contribution. It's looking good.
We need to add a link to the new notebook in `docs/sphinx/using/tutorials.rst` to fix the CI pipeline.

**Comment by 1tnguyen** [comment]
FYI: the notebook validation failure should be fixed by https://github.com/NVIDIA/cuda-quantum/commit/6328008dc5d27b025a7c443df6db19c88354c6c8 (handle `pip install` in the notebook content during validation)
We'll need to wait for that PR to be merged before merging this one.

**Review by bettinaheim — APPROVED** [review]
I'll defer to Thien and Justin's judgement regarding the error mitigation content (discussed/confirmed offline that all looks good)

**Pull request merged**
